### PR TITLE
fix(embed): fix server error due to breaking change on flask-login

### LINKED ISSUE #21987: Embed dashboard on v2.0 generates the following error and throws a 500 internal server error
Error log below for reference

`'LoginManager' object has no attribute 'reload_user'
Traceback (most recent call last):
  File "/root/superset/lib/python3.8/site-packages/flask/app.py", line 1523, in full_dispatch_request
    rv = self.dispatch_request()
  File "/root/superset/lib/python3.8/site-packages/flask/app.py", line 1509, in dispatch_request
    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)
  File "/root/superset/lib/python3.8/site-packages/superset/utils/log.py", line 243, in wrapper
    value = f(*args, add_extra_log_payload=log, **kwargs)
  File "/root/superset/lib/python3.8/site-packages/superset/embedded/view.py", line 60, in embedded
    login_manager.reload_user(AnonymousUserMixin())
AttributeError: 'LoginManager' object has no attribute 'reload_user'
2022-11-01 11:35:46,896:ERROR:superset.views.base:'LoginManager' object has no attribute 'reload_user'
Traceback (most recent call last):
  File "/root/superset/lib/python3.8/site-packages/flask/app.py", line 1523, in full_dispatch_request
    rv = self.dispatch_request()
  File "/root/superset/lib/python3.8/site-packages/flask/app.py", line 1509, in dispatch_request
    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)
  File "/root/superset/lib/python3.8/site-packages/superset/utils/log.py", line 243, in wrapper
    value = f(*args, add_extra_log_payload=log, **kwargs)
  File "/root/superset/lib/python3.8/site-packages/superset/embedded/view.py", line 60, in embedded
    login_manager.reload_user(AnonymousUserMixin())
AttributeError: 'LoginManager' object has no attribute 'reload_user'`

WhatsApp Image 2022-11-01 at 12 10 26 AM

### PR DESCRIPTION
**Pull request by Usiel**
### SUMMARY

Due to a breaking change in Flask-Login (https://github.com/maxcountryman/flask-login/pull/378) the code for logging in an `AnonymousUser` breaks for dashboard embedding. 

Unfortunately, Flask-Login not only renames the method we need, but also makes it quasi-private. We can switch to a different public util function Flask-Login offers since at least version 0.3.0. In all versions I checked it essentially executes the same steps as `reload_user(...)` did (it additionally signals the login event internally, which shouldn't cause issues).

Fixes https://github.com/apache/superset/issues/21987, https://github.com/apache/superset/issues/21146

### TESTING INSTRUCTIONS

For any Flask-Login>=0.3.0,<0.7.0:
1. Enable `EMBEDDED_SUPERSET` in `superset/config.py`
2. Run Superset locally
3. Go to any dashboard, get an embed UUID (see https://github.com/apache/superset/pull/19364 for detailed instructions)
4. Execute the following
```
curl "http://localhost:8088/embedded/<embed_uuid>"
```
**Expected**: 200 OK with some valid HTML
(Before we would fail with a 500 for any Flask-Login>=0.5.0)

### ADDITIONAL INFORMATION

- [x] Has associated issue: https://github.com/apache/superset/issues/21987
- [x] Required feature flags: EMBEDDED_SUPERSET
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

cc @suddjian: Thanks for building this! Tagging you for better visibility and in case I missed anything. Btw, I noticed that `g.user` and the `user` on the `ctx` is actually set to an `AnonymousUser` instance before doing anything in the endpoint; I believe that's some Flask-Login fallback logic. I'm assuming there is some circumstance, some case, where this might not happen and that is why we added the bit of code previously? (That's why I propose to replace it instead of removing it in this PR.)

### CODE REVIEW

**Comment by adimyth** [comment]
Has this been merged & released?

**Comment by pythys** [comment]
I applied the patch to my system and built the docker images, tried to access the dashboards and got instead a new error as attached. I'm not sure if I made a mistake though.

**Comment by pythys** [comment]
ok taking back my comment above, this is not related to this PR, but rather to other security settings. I confirm that the patch is working and fixing my issue on latest code base

**Comment by Usiel** [comment]
> ok taking back my comment above, this is not related to this PR, but rather to other security settings. I confirm that the patch is working and fixing my issue on latest code base

Thanks for verifying the fix @pythys!

> Has this been merged & released?

Not yet @adimyth. Let me bump this PR on the community Slack.

**Comment by adimyth** [comment]
Thanks very much, @Usiel Let me know when you bump this PR in community slack, I'll follow the thread & provide support if needed

**Comment by villebro** [comment]
This appears to be a regression from #22355 . Ping @EugeneTorap , can you check this?

**Comment by EugeneTorap** [comment]
Yes, it's regression bug from my PR because we don't have unit tests for the logic.
In order to prevent the regression bug we need to cover everything with unit tests.

**Comment by EugeneTorap** [comment]
@Usiel Can you add the unit tests for it?

**Comment by villebro** [comment]
> Yes, it's regression bug from my PR because we don't have unit tests for the logic.
> In order to prevent the regression bug in the future we need to cover everything with unit tests.

@EugeneTorap looking more closely it seems it's the missing pinning that was the problem. So we should have restricted to `<0.5` before, and now we should probably have `<0.7` to avoid similar breakage going forward.

**Comment by EugeneTorap** [comment]
@villebro We use only ´"flask-login==0.6.0"´ in setup.py because 0.6.2 has breaking changes for superset.

**Comment by Usiel** [comment]
> @Usiel Can you add the unit tests for it?

Sure, I added an integration test (plus some other cases), which fails with the previous implementation and `Flask-Login>0.4`.

**Comment by EugeneTorap** [comment]
@Usiel Thank you!
@dpgaspar Can you review the PR?

File: tests/integration_tests/embedded/view_tests.py [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+from superset.embedded.dao import EmbeddedDAO
+from superset.models.dashboard import Dashboard
+from tests.integration_tests.base_tests import SupersetTestCase
+from tests.integration_tests.fixtures.birth_names_dashboard import (
+    load_birth_names_dashboard_with_slices,
+    load_birth_names_data,
+)
+
+
+class TestEmbeddedDashboardView(SupersetTestCase):
```
Comment (villebro) [inline tests/integration_tests/embedded/view_tests.py:31]: Nit: we try to steer in the direction of functional tests using `pytest`, but I'm ok with these if you're not familiar with them.

**Review by villebro — APPROVED** [review]
Thanks so much for this fix + the added tests, a much needed improvement to ensure embedded is properly maintained going forward 👍 One minor nit, but I'm happy merging as-is.

**Comment by adimyth** [comment]
Looking forward to this PR being merged & being shipped in the next release!

**Comment by cwegener** [comment]
> we should have restricted to `<0.5` before, and now we should probably have `<0.7` to avoid similar breakage going forward.

If assuming SemVer 2.0, this is not true, as releases in Major version `0.y.z` SHOULD NOT have their Public API considered stable.

File: tests/integration_tests/embedded/view_tests.py (hunk 1 above)
Comment (Usiel) [reply tests/integration_tests/embedded/view_tests.py]: Not super familiar, but I tried my best in https://github.com/apache/superset/pull/22462/commits/0717fb74b601d48c119295775ae92802ac46b5e0 :)

**Comment by villebro** [comment]
> > we should have restricted to `<0.5` before, and now we should probably have `<0.7` to avoid similar breakage going forward.
> 
> If assuming SemVer 2.0, this is not true, as releases in Major version `0.y.z` SHOULD NOT have their Public API considered stable.

Strictly speaking this is true, but in my experience most patch releases in the 0 major version tend to be API stable.

File: tests/integration_tests/embedded/test_view.py [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+    from flask.testing import FlaskClient
+
+
+@pytest.mark.usefixtures("load_birth_names_dashboard_with_slices")
+@mock.patch.dict(
+    "superset.extensions.feature_flag_manager._feature_flags",
+    EMBEDDED_SUPERSET=True,
+)
+def test_get_embedded_dashboard(client: "FlaskClient[Any]"):
```
Comment (villebro) [inline tests/integration_tests/embedded/test_view.py:42]: nit: you can remove the quotes if you add `from __future__ import annotations` before the other imports:
```suggestion
def test_get_embedded_dashboard(client: FlaskClient[Any]):
```

File: tests/integration_tests/embedded/view_tests.py (hunk 1 above)
Comment (villebro) [reply tests/integration_tests/embedded/view_tests.py]: @Usiel looks great! One minor nit to remove redundant quotes on the type annotation.

File: tests/integration_tests/embedded/test_view.py (hunk 2 above)
Comment (Usiel) [reply tests/integration_tests/embedded/test_view.py]: Oh interesting, pushed a commit

**Pull request merged**